### PR TITLE
Don't default parser to babylon

### LINKED ISSUE #2884: Consider removing default parser / Skip processing unsupported file types
Currently when

* Prettier is invoked from the CLI, and
* The file being formatted doesn't match any of the extensions supported, and
* No `--parser` option is passed, and
* No `parser` is in a prettier config file,

then it will fall back to the babylon parser and treat the file as JavaScript. This made sense when prettier only supported JS (and then TypeScript), but now that it supports vastly different languages perhaps it is time to challenge that default.

In #2882, we discovered this can have negative repercussions when used with `--write **/*`, modifying files inside of a `.git` directory.

If we were to change the default from JavaScript to a no-op, these kind of issues would go away, and we'd also have less issues reported along the lines of "My CSS doesn't parse", when they're accidentally using the babylon parser.

It is already possible to associate other extensions with a parser in `.prettierrc`.

```json
{
  "overrides": [{
     "files": "*.mjs",
     "options": { "parser": "babylon" }
  }]
}
```

Thoughts @lydell @vjeux?

### PR DESCRIPTION
**Pull request by duailibe**
Closes #2884

This takes a different approach from #4426 to make it more explicit a parser wasn't specified or inferred (returns normalized opts as `null` and `format()` bails formatting).

### CODE REVIEW

File: src/main/options.js [hunk 1]
```diff
@@ -29,30 +30,28 @@ function normalize(options, opts) {
...
-      rawOptions.filepath &&
-      (!rawOptions.parser || rawOptions.parser === defaults.parser)
-    ) {
-      const inferredParser = inferParser(
-        rawOptions.filepath,
-        rawOptions.plugins
+  if (!rawOptions.parser) {
+    const logger = (opts && opts.logger) || console;
+    if (!rawOptions.filepath) {
+      logger.warn("Skipping file with unknown filepath");
```
Comment (lydell) [inline src/main/options.js]: What is an unknown filepath?

Comment (duailibe) [reply src/main/options.js]: @lydell Use stdin without `--stdin-filepath` or `prettier.format` without `{ filepath }`

Comment (duailibe) [reply src/main/options.js]: At first I decided to throw an error but I figured that could be a little too harsh

**Comment by lydell** [comment]
I fetched your branch and played around with it a little, and it seems nice.

Minor logging glitch:

```
❯ ./bin/prettier.js LICENSE index.js --write
LICENSESkipping file with unknown extension: LICENSE
LICENSE 2ms
index.js 59ms
```

Question about stdout:

```
❯ ./bin/prettier.js LICENSE
```

Currently the above print LICENSE as-is to stdout. Is it better to not print it to stdout at all?

File: src/main/options.js (hunk 1 above)
Comment (lydell) [reply src/main/options.js]: I see. I wanted to come up with a better message, but it's hard.

File: src/main/core-options.js [hunk 2]
```diff
@@ -85,7 +85,7 @@ const options = {
     since: "0.0.10",
     category: CATEGORY_GLOBAL,
     type: "choice",
-    default: "babylon",
+    default: undefined,
```
Comment (suchipi) [inline src/main/core-options.js]: For the support info, you need to specify that the default varies across versions; see https://github.com/prettier/prettier/pull/4426#discussion_r186282273

**Comment by duailibe** [comment]
> Currently the above print LICENSE as-is to stdout. Is it better to not print it to stdout at all? Similar: What should `prettier.format("foo")` return?

Good questions, I'm not sure. I'm comfortable with throwing an error in the latter, but I'm not sure what we should do about `prettier LICENSE`

**Comment by duailibe** [comment]
So I tried to improve the logging experience a bit but there are too much cases to cover 😅 

This code is pretty complex

**Comment by duailibe** [comment]
I *think* I covered most use cases now in `infer-parser.js`; should be easier to reason about the behavior now

File: tests_integration/__tests__/infer-parser.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  });
+
+  describe("multiple files", () => {
+    runPrettier("cli/infer-parser/", ["--write", "*"]).test({
+      status: 2
+    });
+  });
+});
+
+describe("--write and --list-different with unknown path and no parser", () => {
```
Comment (duailibe) [inline tests_integration/__tests__/infer-parser.js]: I have no idea what this means.. we really need to revamp the CLI

File: src/main/options.js (hunk 1 above)
Comment (lydell) [reply src/main/options.js]: Really like the updated message, good job!

**Comment by suchipi** [comment]
🎉

**Pull request merged**